### PR TITLE
Start adding support for FIPS sepcific features

### LINKED ISSUE #46: Add support for the upcoming 3.2 Spec certification indicators
PKCS#11 3.2 will introduce a set of indicators to report whether a cryptographic operation was performed in accordance with approved certification modes.
This is necessary specifically for FIPS-140-3 certifications.

### PR DESCRIPTION
**Pull request by simo5**
Adds attributes needed to deal with FIPS Indicators.

This will only be standardized in PKCS#11 3.2 so for now this is a Draft PR.

Resolves #46

### CODE REVIEW

**Comment by simo5** [comment]
coverity found an issue:
```
*** CID 496758:  Memory - illegal accesses  (OVERRUN)
/src/provider.c: 1203 in p11prov_query_operation()
1197     p11prov_query_operation(void *provctx, int operation_id, int *no_cache)
1198     {
1199         P11PROV_CTX *ctx = (P11PROV_CTX *)provctx;
1200    
1201         if (operation_id > OSSL_OP__HIGHEST
1202             || ctx->blocked_ops[operation_id] == true) {
>>>     CID 496758:  Memory - illegal accesses  (OVERRUN)
>>>     Overrunning array "p11prov_block_ops_names" of 23 8-byte elements at element index 23 (byte offset 191) using index "operation_id" (which evaluates to 23).
1203             P11PROV_debug("Blocked operation: %s (%d)",
1204                           p11prov_block_ops_names[operation_id], operation_id);
1205             *no_cache = 0;
1206             return NULL;
1207         }
1208 
```   

Will adjust the debug statement to not overflow the struct.

File: src/fips/indicators.rs [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+        data.attributes.push(attr_element!(CKA_VALIDATION_AUTHORITY_TYPE; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_ulong; val 0));
+        data.attributes.push(attr_element!(CKA_VALIDATION_COUNTRY; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_string; val String::new()));
+        data.attributes.push(attr_element!(CKA_VALIDATION_CERTIFICATE_IDENTIFIER; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_string; val String::new()));
+        data.attributes.push(attr_element!(CKA_VALIDATION_CERTIFICATE_URI; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_string; val String::new()));
+        data.attributes.push(attr_element!(CKA_VALIDATION_VENDOR_URI; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_string; val String::new()));
+        data.attributes.push(attr_element!(CKA_VALIDATION_PROFILE; OAFlags::AlwaysRequired | OAFlags::NeverSettable | OAFlags::Unchangeable; from_string; val String::new()));
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:44]: I guess no for macros?
No idea.

File: src/fips/indicators.rs [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+    ($s1:expr) => {
+        ([btb!($s1), 0, 0, 0], (0, 0))
+    };
+    ($s1:expr, $s2: expr) => {
+        ([btb!($s1), btb!($s2), 0, 0], (0, 0))
+    };
+    ($s1:expr, $s2: expr, $s3:expr) => {
+        ([btb!($s1), btb!($s2), btb!($s3), 0], (0, 0))
+    };
+    ($s1:expr, $s2: expr, $s3:expr, $s4:expr) => {
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:136]: interesting, I guess a corner case missed by the formatter

File: src/fips/indicators.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:44]: Manually reformatting

File: src/fips/indicators.rs [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+    obj.set_attr(attribute::from_string(
+        CKA_VALIDATION_PROFILE,
+        String::from(""),
+    ))?;
+
+    /* generate a unique id */
+    obj.generate_unique();
+
+    /* invalid session handle will prevent it from being removed when
+     * session objects are cleared on session closings */
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:114]: Not sure, but I will indeed add a comment there

File: src/fips/indicators.rs [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+        FipsKeyType {
+            keytype: CKK_SHA3_512_HMAC,
+            operations: CKF_SIGN | CKF_VERIFY | CKF_DERIVE,
+            sizes: step!(512),
+        },
+        FipsKeyType {
+            keytype: CKK_EC_EDWARDS,
+            operations: CKF_SIGN | CKF_VERIFY,
+            sizes: step!(256, 456),
+        },
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:300]: pretty sure I got this from nss code, and should be correct.
It is ok if we do not support Ed448 for now, this list will have to be thoroughly reviewed and adjusted by anyone that intends to certify.

File: src/fips/indicators.rs [hunk 5]
```diff
@@ -0,0 +1,1057 @@
...
+            operations: CKF_SIGN | CKF_VERIFY,
+            restrictions: [restrict!(CKK_RSA), restrict!()],
+            genflags: 0,
+        },
+        /* ECC */
+        FipsMechanism {
+            mechanism: CKM_EC_KEY_PAIR_GEN,
+            operations: CKF_GENERATE_KEY_PAIR,
+            restrictions: [restrict!(CKK_EC), restrict!()],
+            genflags: CKF_SIGN | CKF_VERIFY,
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:445]: good catch

File: src/fips/indicators.rs [hunk 6]
```diff
@@ -0,0 +1,1057 @@
...
+            mechanism: CKM_SHA3_512_RSA_PKCS_PSS,
+            operations: CKF_SIGN | CKF_VERIFY,
+            restrictions: [restrict!(CKK_RSA), restrict!()],
+            genflags: 0,
+        },
+        /* ECC */
+        FipsMechanism {
+            mechanism: CKM_EC_KEY_PAIR_GEN,
+            operations: CKF_GENERATE_KEY_PAIR,
+            restrictions: [restrict!(CKK_EC), restrict!()],
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs:444]: CKK_EC_EDWARDS is still TODO in this code see check_key()

File: src/lib.rs [hunk 7]
```diff
@@ -2384,45 +2461,37 @@ pub extern "C" fn C_GetInterface(
...
-            return CKR_ARGUMENTS_BAD;
-        }
-    }
-    if !version.is_null() {
-        unsafe {
-            ver.major = (*version).major;
-            ver.minor = (*version).minor;
-        }
+    if version.is_null() {
+        return CKR_ARGUMENTS_BAD;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/lib.rs:2468]: uhmmm, ok, let me fix this

[the remark this answers is not in the document]
Comment (simo5) [reply src/lib.rs]: should do the right thing now

File: src/fips/indicators.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/fips/indicators.rs:332]: Not according to the PKCS#11 specs -- I knew I had to take it from somewhere, but it took me some time to find it:

> For this mechanism, the ulMinKeySize and ulMaxKeySize fields of the CK_MECHANISM_INFO structure specify the minimum and maximum supported number of bits in the field sizes, respectively.  For this mechanism, the only allowed values are 255 and 448 as RFC 8032and RFC 8410 only define curves of these two sizes.  A Cryptoki implementation may support one or both of these curves and should set the ulMinKeySize and ulMaxKeySize fields accordingly.

https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html#_Toc30061191

[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs]: uh I had made the change, and somehow it got reverted when I rebased ...

[the remark this answers is not in the document]
Comment (simo5) [reply src/fips/indicators.rs]: uh no, it is actually fixed, the two values are now 255, 448

[the remark this answers is not in the document]
Comment (Jakuje) [reply src/fips/indicators.rs]: I probably missed that. Thanks for double-checking.

**Review by Jakuje — APPROVED** [review]

**Pull request merged**